### PR TITLE
gRPC: fix goroutine leak if client stream CloseSend is not called.

### PR DESCRIPTION
**Pull request by vadimberezniker**
Outside of a unidirectional Client->Server stream there's no requirement
that CloseSend must be called (and it doesn't make much sense for a
Server->Client stream).

If CloseSend is not called, the RPC span will not be ended and a
goroutine will be leaked.

We saw a large number of goroutines leaked when we enabled gRPC
instrumentation in our service.

Resolves https://github.com/open-telemetry/opentelemetry-go-contrib/issues/818

### CODE REVIEW

**Review by Aneurysm9 — APPROVED** [review]

**Comment by vadimberezniker** [comment]
Friendly ping :)

File: instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go [hunk 1]
```diff
@@ -208,24 +200,15 @@ func wrapClientStream(s grpc.ClientStream, desc *grpc.StreamDesc) *clientStream
...
-		// Both streams have to be closed
-		state := byte(0)
-
 		for event := range events {
 			switch event.Type {
-			case closeEvent:
-				state |= clientClosedState
 			case receiveEndEvent:
-				state |= receiveEndedState
+				finished <- nil
```
Comment (coryb) [inline instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:206]: I was running into the same issues and after applying this PR I was still seeing some leaks.  I think this fixes some of the issues, but I think there is still an edge case where the client context is cancelled in which case we may not receive any more events.  I patched this loop to select on the `s.Context.Done()` along with `events` and all my leaks have gone away.  This is what my loop looks like:
```
		for {
			select {
			case <-s.Context().Done():
				finished <- s.Context().Err()
				return
			case event := <-events:
				switch event.Type {
				case receiveEndEvent:
					finished <- nil
					return
				case errorEvent:
					finished <- event.Err
					return
				}
			}
		}
```

Comment (vadimberezniker) [reply instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:206]: Please see https://github.com/open-telemetry/opentelemetry-go-contrib/pull/840

**Review by tonistiigi — APPROVED** [review]

**Comment by vadimberezniker** [comment]
Ping. This PR has been open for 2 weeks now.

**Review by XSAM — APPROVED** [review]
💯

**Comment by vadimberezniker** [comment]
Thank you! Any blockers for merging this?

**Comment by vadimberezniker** [comment]
What's left?

**Review by MrAlias — APPROVED** [review]

**Pull request merged**